### PR TITLE
[MAINTENANCE] Improve type checking for Expectations with atomic renderers leveraging `RendererConfiguration`

### PR DESCRIPTION
**Pull request by NathanFarmer**
Changes proposed in this pull request:
- GREAT-1478
- Updated existing renderers that use `RendererConfiguration` due to static type check errors
- Fixed additional Expectation-focused static type check errors

### Definition of Done
- [X] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/docs/contributing/style_guides/code_style)
- [X] I have performed a [self-review](https://docs.greatexpectations.io/docs/contributing/contributing_checklist) of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [X] I have added [unit tests](https://docs.greatexpectations.io/docs/contributing/contributing_test#writing-unit-and-integration-tests) where applicable and made sure that new and existing tests are passing.
- [X] I have run any local integration tests and made sure that nothing is broken.

[GREAT-1478]: https://greatexpectations.atlassian.net/browse/GREAT-1478?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -86,7 +86,7 @@ exclude = [
     'execution_engine/split_and_sample/sparkdf_data_splitter\.py',  # 4
     'execution_engine/split_and_sample/sqlalchemy_data_sampler\.py',  # 10
     'execution_engine/split_and_sample/sqlalchemy_data_splitter\.py',  # 22
-    'expectations/core/expect_column_', # 214
+    'expectations/core/expect_column_', # 147
     'expectations/core/expect_compound_columns_to_be_unique\.py', # 3
```
Comment (Kilo59) [inline pyproject.toml:89]: If it makes things easier, you could replace this with a more specific regex, rather than trying to fix all `214` warnings at once.

Comment (NathanFarmer) [reply pyproject.toml:89]: If it works for you, I would prefer to do this in the next PR, since I expect this inventory to change drastically from this PR to the next.

File: contrib/great_expectations_semantic_types_expectations/great_expectations_semantic_types_expectations/expectations/expect_column_values_to_be_valid_price.py [hunk 2]
```diff
@@ -92,7 +92,7 @@ class ExpectColumnValuesToBeValidPrice(ColumnMapExpectation):
     default_kwarg_values = {}
 
     def validate_configuration(
-        self, configuration: Optional[ExpectationConfiguration]
+        self, configuration: Optional[ExpectationConfiguration] = None
```
Comment (cdkini) [inline contrib/great_expectations_semantic_types_expectations/great_expectations_semantic_types_expectations/expectations/expect_column_values_to_be_valid_price.py:95]: Why this change? Is there ever a situation where this is called without an input config?

File: tests/integration/docusaurus/expectations/creating_custom_expectations/expect_multicolumn_values_to_be_multiples_of_three.py [hunk 3]
```diff
@@ -121,15 +121,12 @@ def validate_configuration(
...
-                    "column_list" in configuration.kwargs
-                ), "column_list must be provided"
-            except AssertionError as e:
-                raise InvalidExpectationConfigurationError(str(e))
+        configuration = configuration or self.configuration
+
+        try:
+            assert "column_list" in configuration.kwargs, "column_list must be provided"
+        except AssertionError as e:
+            raise InvalidExpectationConfigurationError(str(e))
```
Comment (cdkini) [inline tests/integration/docusaurus/expectations/creating_custom_expectations/expect_multicolumn_values_to_be_multiples_of_three.py:129]: Instead of an assert`, can't we just do:
```
if "column_list" not in configuration.kwargs:
    raise InvalidExpectationConfigurationError("column_list_must_be_provided")
```

File: great_expectations/expectations/expectation.py [hunk 4]
```diff
@@ -2413,10 +2413,10 @@ def validate_configuration(
 
 
 class ColumnMapExpectation(TableExpectation, ABC):
-    map_metric = None
+    map_metric = ""
```
Comment (cdkini) [inline great_expectations/expectations/expectation.py]: This change won't be an issue right? Seems to be a base for a lot of classes

File: great_expectations/expectations/expectation.py [hunk 5]
```diff
@@ -2285,7 +2285,7 @@ def _validate_metric_value_between(
         else:
             below_max = True
 
-        success = above_min and below_max
+        success = bool(above_min and below_max)
```
Comment (cdkini) [inline great_expectations/expectations/expectation.py:2288]: Is this just for readability or was it required by a linter? I think its assumed to be a bool based on the conditional no?

File: contrib/great_expectations_semantic_types_expectations/great_expectations_semantic_types_expectations/expectations/expect_column_values_to_be_valid_price.py (hunk 2 above)
Comment (NathanFarmer) [inline contrib/great_expectations_semantic_types_expectations/great_expectations_semantic_types_expectations/expectations/expect_column_values_to_be_valid_price.py:95]: In the entire class hierarchy this is what the signature of this method looks like, except for sometimes, at the child Expectation-level. The super always gets called at the beginning of this method and at every level it checks if `configuration` is `None` and if so, use `self.configuration`.

File: great_expectations/expectations/core/expect_column_values_to_not_be_in_set.py [hunk 6]
```diff
@@ -143,16 +142,16 @@ def _prescriptive_template(
...
         renderer_configuration: RendererConfiguration,
     ) -> RendererConfiguration:
-        add_param_args = (
+        add_param_args: List[AddParamArgs] = [
             ("column", RendererValueType.STRING),
             ("value_set", RendererValueType.ARRAY),
             ("mostly", RendererValueType.NUMBER),
             ("parse_strings_as_datetimes", RendererValueType.BOOLEAN),
-        )
+        ]
```
Comment (Kilo59) [inline great_expectations/expectations/core/expect_column_values_to_not_be_in_set.py:150]: All things being equal, I think. a tuple would be better since it's immutable.

File: contrib/great_expectations_semantic_types_expectations/great_expectations_semantic_types_expectations/expectations/expect_column_values_to_be_valid_price.py (hunk 2 above)
Comment (NathanFarmer) [reply contrib/great_expectations_semantic_types_expectations/great_expectations_semantic_types_expectations/expectations/expect_column_values_to_be_valid_price.py:95]: I'm not aware of the workflow where this could be none, but I think the pattern should remain the same, especially given the `Optional` annotation.

File: great_expectations/expectations/core/expect_column_value_lengths_to_be_between.py [hunk 7]
```diff
@@ -278,30 +276,30 @@ def _prescriptive_template(
...
-        add_param_args = (
+        add_param_args: List[AddParamArgs] = [
             ("column", RendererValueType.STRING),
             ("min_value", [RendererValueType.NUMBER, RendererValueType.DATETIME]),
             ("max_value", [RendererValueType.NUMBER, RendererValueType.DATETIME]),
             ("mostly", RendererValueType.NUMBER),
             ("strict_min", RendererValueType.BOOLEAN),
             ("strict_max", RendererValueType.BOOLEAN),
-        )
+        ]
```
Comment (Kilo59) [inline great_expectations/expectations/core/expect_column_value_lengths_to_be_between.py:286]: Similar question as below. Is there a reason we need a `list`? Tuples are immutable and use less memory.

File: great_expectations/expectations/core/expect_column_unique_value_count_to_be_between.py [hunk 8]
```diff
@@ -213,30 +213,30 @@ def _prescriptive_template(
         cls,
         renderer_configuration: RendererConfiguration,
     ) -> RendererConfiguration:
-        add_param_args = (
+        add_param_args: List[AddParamArgs] = [
```
Comment (Kilo59) [inline great_expectations/expectations/core/expect_column_unique_value_count_to_be_between.py:216]: tuple vs list?

File: great_expectations/expectations/core/expect_column_stdev_to_be_between.py [hunk 9]
```diff
@@ -214,30 +214,30 @@ def _prescriptive_template(
         cls,
         renderer_configuration: RendererConfiguration,
     ) -> RendererConfiguration:
-        add_param_args = (
+        add_param_args: List[AddParamArgs] = [
```
Comment (Kilo59) [inline great_expectations/expectations/core/expect_column_stdev_to_be_between.py:217]: tuple vs list?

File: great_expectations/expectations/core/expect_column_values_to_be_unique.py [hunk 10]
```diff
@@ -104,64 +112,35 @@ def validate_configuration(
...
-            },
-            "condition_parser": {
-                "schema": {"type": "string"},
-                "value": params.get("condition_parser"),
-            },
-        }
+        renderer_configuration: RendererConfiguration,
+    ) -> RendererConfiguration:
+        add_param_args: List[AddParamArgs] = [
+            ("column", RendererValueType.STRING),
```
Comment (cdkini) [inline great_expectations/expectations/core/expect_column_values_to_be_unique.py:120]: Could the first value of this tuple become an enum as well? Not necessary in this PR but just asking!

File: docs/guides/expectations/creating_custom_expectations/how_to_create_custom_column_aggregate_expectations.md [hunk 11]
```diff
@@ -322,7 +322,7 @@ This guide will leave you with a Custom Expectation sufficient for [contribution
 
 If you plan to contribute your Expectation to the public open source project, you should update the `library_metadata` object before submitting your [Pull Request](https://github.com/great-expectations/great_expectations/pulls). For example:
 
-```python file=../../../../examples/expectations/column_aggregate_expectation_template.py#L105-L110
+```python file=../../../../examples/expectations/column_aggregate_expectation_template.py#L104-L109
```
Comment (cdkini) [inline docs/guides/expectations/creating_custom_expectations/how_to_create_custom_column_aggregate_expectations.md:325]: Appreciate all the changes around here - I know these line number refs are a bit annoying

**Review by Kilo59 — APPROVED** [review]
Looks good to me.
I would consider reverting the `tuple` -> `list` changes unless there's an obvious benefit that outweighs the memory and immutable benefit of `tuples.

File: great_expectations/expectations/expectation.py (hunk 4 above)
Comment (Kilo59) [inline great_expectations/expectations/expectation.py]:
```suggestion
    map_metric: str = ""
```

Should we say this must be a string?

Comment (NathanFarmer) [reply great_expectations/expectations/expectation.py]: I ended up changing this to `map_metric: Optional[str] = None` to make the type checker happy, because you were right that this is `None` vs `str` specific logic in this class.

File: great_expectations/expectations/expectation.py (hunk 5 above)
Comment (NathanFarmer) [reply great_expectations/expectations/expectation.py:2288]: Nice catch! This was an unrelated bugfix for @alexsherstinsky due to issues when the variables are `np.bool`.

File: great_expectations/expectations/expectation.py (hunk 4 above)
Comment (NathanFarmer) [reply great_expectations/expectations/expectation.py]: I ended up needing to revert this due to some `isinstance` checks on `str` vs `None` with error messages in this class.

File: great_expectations/expectations/core/expect_column_values_to_be_unique.py (hunk 10 above)
Comment (NathanFarmer) [reply great_expectations/expectations/core/expect_column_values_to_be_unique.py:120]: I don't think it would work like that. These strings usually map the Expectation kwargs and are one of the only Expectation-specific things in the renderer methods.

File: great_expectations/expectations/core/expect_column_values_to_not_be_in_set.py (hunk 6 above)
Comment (NathanFarmer) [reply great_expectations/expectations/core/expect_column_values_to_not_be_in_set.py:150]: I think I made this change because I forgot you can annotate a tuple of unknown length `Tuple[AddParamArgs, ...]`. I will make this change in the next PR since I need to change some of it as well.

File: tests/integration/docusaurus/expectations/creating_custom_expectations/expect_multicolumn_values_to_be_multiples_of_three.py (hunk 3 above)
Comment (NathanFarmer) [reply tests/integration/docusaurus/expectations/creating_custom_expectations/expect_multicolumn_values_to_be_multiples_of_three.py:129]: I would prefer to punt on this since this not really in the scope of this PR and is repeated many places. Also, once you change it in those places, you have to update all of the line numbers in the docs again!

File: great_expectations/expectations/expectation.py (hunk 4 above)
Comment (Kilo59) [reply great_expectations/expectations/expectation.py]: That sounds like incorrect logic.
An empty string shouldn't be a valid `map_metric`??

Comment (Kilo59) [reply great_expectations/expectations/expectation.py]: Seems like that `str` vs `None` logic might be wrong.
I would expect an empty string would be invalid.
